proposed/2021/DotNetAudit.md

```diff
@@ -0,0 +1,296 @@
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+.NET developers manage their dependencies on a daily basis. However, it's become more and more difficult to know exactly what dependencies in a software supply chain might be vulnerable, deprecated, or even outdated. All of these scenarios require a developers direct attention to resolve. Even knowing about a dependency that falls into one of these categories, there is then the question about what to do with said dependency such as:
+
+- Removing it
+- Replacing it
+- Updating it
+- Ignoring it
+
+The goal of this proposal is to provide developers a single command to audit their dependencies, resolve them, and alert the user to manually take action towards a resolution if the tooling cannot.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+#### dotnet audit
+
+The `dotnet audit` command will scan your project and solution's dependencies and look for known vulnerabilities, deprecations, and outdated packages. If any vulnerabilities, deprecations, or outdated packages are found, a remediation will be calculated to then `fix` which will be applied to the resulting package graph.
+
+Example:
+
+![dotnet audit](/meta/resources/dotnetaudit/dotnetaudit.png)
+
+```
+$ dotnet audit
+
+Fetching package metadata from: 'https://api.nuget.org/v3/index.json'
+Loaded 23 security advisories from 'https://api.nuget.org/v3/index.json'
+Scanning ContosoUniversity.sln (36 NuGet dependencies)
+
+error: Vulnernable packages found!
+[net5.0]:
+Top-level Package	    Requested	Resolved	Severity	Advisory URL
+> UmbracoForms		    8.4.1		8.4.1		Moderate	https://github.com/advisories/GHSA-8m73-w2r2-6xxj
+
+Transitive Package		Resolved	Severity	Advisory URL
+> Microsoft.Data.OData	5.2.0		Moderate	https://github.com/advisories/GHSA-mv2r-q4g5-j8q5
+
+Found 1 top-level Moderate severity vulnerability & 1 transtive Moderate severity vulnerability package(s) in 36 scanned packages.
+
+Run 'dotnet audit fix' to fix them.
+
+warning: Deprecated packages found!
+[net5.0]:
+Top-level Package						Requested	Resolved	Reason(s)	Alternative
+> anurse.testing.TestDeprecatedPackage	1.0.0		1.0.0		Legacy		Microsoft.AspNetCore.Mvc > 0.0.0
+
+Found 1 top-level Legacy deprecated package(s) in 36 scanned packages.
+
+Run 'dotnet audit fix' to fix them.
+
+warning: Outdated packages found!
+[net5.0]:
+Top-level Package						            Resolved	Latest
+> anurse.testing.TestDeprecatedPackage	            1.0.0		2.0.0
+> UmbracoForms							            8.4.1	    8.7.1
+
+Found 2 top-level outdated package(s) in 36 scanned packages.
```

Comment:
Would this report transitive packages that are outdated? Would users have to hoist outdated transitive dependencies to update them to the latest version? That may be a frustrating experience for customers.

Perhaps we should only report outdated top-level dependencies. If so, consider adding a note on that limitation somewhere.

Reply:
It currently follows the outdated experience which doesn't include transitives. I'm also not so sure what value seeing outdated transitives would have unless one wanted to promote those explicitly. Will have to circle back here.